tests/puzzle/test_solver.py

```diff
@@ -4707,6 +4707,47 @@ def test_solver_resolves_duplicate_dependency_in_extra(
     )
 
 
+@pytest.mark.parametrize("with_extra", [False, True])
+def test_solver_resolves_duplicate_dependency_in_root_extra(
+    package: ProjectPackage,
+    pool: RepositoryPool,
+    repo: Repository,
+    io: NullIO,
+    with_extra: bool,
+) -> None:
+    """
+    Without extras, a newer version of A can be chosen than with root extras.
+    """
```

Comment:
This is only a theoretical issue. In other words, the setup of the test cannot happen when invoking poetry.

First of all, you cannot create this issue with Poetry 1.8 because if you try something like:

```
[tool.poetry.dependencies]
python = "^3.9"
tomli = [
    { version = ">=1" },
    { version = "^1", optional = true },
]

[tool.poetry.extras]
foo = [ "tomli" ]
```

... both dependencies are in the extra. However, with the main branch (by the way, please rebase your branch when it suits you) and PEP 621 support you can define:

```
[project]
# ...
requires-python = ">=3.9"
dependencies = [
    "tomli>=1",
]

[project.optional-dependencies]
foo = [
    "tomli>=1,<2",
]
```

When you debug this, you will notice that the tomli dependency from the extra has no `extra == "foo"` marker. Generally speaking, dependencies of root extras do not have the extra in the marker but only the attribute `_in_extras` set. Thus, a more realistic test will look like that (and does not fail):

```
def test_solver_resolves_duplicate_dependency_in_root_extra(
    package: ProjectPackage,
    pool: RepositoryPool,
    repo: Repository,
    io: NullIO,
) -> None:
    """
    Without extras, a newer version of A could be chosen than with root extras.
    However, only the older version is chosen.
    """
    package_a1 = get_package("A", "1.0")
    package_a2 = get_package("A", "2.0")

    dep = get_dependency("A", ">=1.0")
    package.add_dependency(dep)

    dep_extra = get_dependency("A", "^1.0", optional=True)
    dep_extra._in_extras = ["foo"]
    package.extras = {canonicalize_name("foo"): [dep_extra]}
    package.add_dependency(dep_extra)

    repo.add_package(package_a1)
    repo.add_package(package_a2)

    solver = Solver(package, pool, [], [], io)
    transaction = solver.solve()

    check_solver_result(transaction, ([{"job": "install", "package": package_a1}]))
```

I only checked this test for now and will take a look at the installer tests later.